app/jobs/prepare_transfer_job.rb

```diff
@@ -1,96 +1,91 @@
 # frozen_string_literal: true
 
 # rubocop:disable Metrics/AbcSize
+# rubocop:disable Metrics/CyclomaticComplexity
 # rubocop:disable Metrics/MethodLength
+# rubocop:disable Metrics/PerceivedComplexity:
 # rubocop:disable Style/MethodCallWithoutArgsParentheses
 
 class PrepareTransferJob < ApplicationJob
   queue_as Atc::Queues::PREPARE_TRANSFER
 
   def perform(source_object_id, enqueue_successor: true)
     source_object = SourceObject.find(source_object_id)
-
     whole_file_crc32c_checksum = nil
-
     pending_transfers = []
 
-    queue_for_gcp = queue_for_gcp?(source_object)
-    if queue_for_aws?(source_object)
+    aws_storage_providers = select_storage_providers(source_object, 'aws')
+    gcp_storage_providers = select_storage_providers(source_object, 'gcp')
+
+    if aws_storage_providers.present?
       # AWS uses whole file checksum for smaller files, and multipart checksum for larger files
       if File.size(source_object.path) < Atc::Constants::DEFAULT_MULTIPART_THRESHOLD
-        whole_file_crc32c_checksum = Digest::CRC32c.file(source_object.path).digest
-        pending_transfers << PendingTransfer.create!(
-          transfer_checksum_algorithm: crc32c_checksum_algorithm(),
-          transfer_checksum_value: whole_file_crc32c_checksum,
-          storage_provider: aws_storage_provider(),
-          source_object: source_object
-        )
+        whole_file_crc32c_checksum ||= Digest::CRC32c.file(source_object.path).digest
+        aws_storage_providers.each do |aws_storage_provider|
+          pending_transfers << PendingTransfer.create!(
+            transfer_checksum_algorithm: crc32c_checksum_algorithm(),
+            transfer_checksum_value: whole_file_crc32c_checksum,
+            storage_provider: aws_storage_provider,
+            source_object: source_object
+          )
+        end
       else
-        checksum_data = Atc::Utils::AwsChecksumUtils.multipart_checksum_for_file(source_object.path,
-                                                                                 calculate_whole_object: queue_for_gcp)
-        whole_file_crc32c_checksum = checksum_data[:binary_checksum_of_whole_file]
-        pending_transfers << PendingTransfer.create!(
-          transfer_checksum_algorithm: crc32c_checksum_algorithm(),
-          transfer_checksum_value: checksum_data[:binary_checksum_of_checksums],
-          transfer_checksum_part_size: checksum_data[:part_size],
-          transfer_checksum_part_count: checksum_data[:num_parts],
-          storage_provider: aws_storage_provider(),
-          source_object: source_object
+        checksum_data ||= Atc::Utils::AwsChecksumUtils.multipart_checksum_for_file(
+          source_object.path,
+          calculate_whole_object: gcp_storage_providers.present?
         )
+        whole_file_crc32c_checksum = checksum_data[:binary_checksum_of_whole_file]
```

Comment:
Is whole_file_crc32c_checksum used in the local scope defined inside lines 32 and 48? It is used beyond line 48, but initialized first on line 53. So line 37 can be removed if I'm not mistaken.
This extra line seems to be in the pre-PR code as well, so I guess it can stay if we want to limit the PR changes.

Reply:
Line 37 is a small optimization that runs under a specific set of conditions: when an aws storage provider is present AND the object size is above the aws multipart threshold AND there's a gcp storage provider present.  In this scenario, we read the underlying file only one time while generating both multipart and whole-file checksums, store both values in that `checksum_data` variable, and then that allows us to skip the calculation on line 53.